core/web/health_controller.go

```diff
@@ -96,13 +99,202 @@ func (hc *HealthController) Health(c *gin.Context) {
 			Output: output,
 		})
 	}
+	slices.SortFunc(checks, func(a, b presenters.Check) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
+
+	switch c.NegotiateFormat(gin.MIMEJSON, gin.MIMEHTML, gin.MIMEPlain) {
+	case gin.MIMEJSON:
+		break // default
+
+	case gin.MIMEHTML:
+		if err := newCheckTree(checks).WriteHTMLTo(c.Writer); err != nil {
+			hc.App.GetLogger().Errorw("Failed to write HTML health report", "err", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+		}
+		return
 
-	if testing.Testing() {
+	case gin.MIMEPlain:
 		slices.SortFunc(checks, func(a, b presenters.Check) int {
 			return cmp.Compare(a.Name, b.Name)
 		})
+
+		if err := writeTextTo(c.Writer, checks); err != nil {
+			hc.App.GetLogger().Errorw("Failed to write plaintext health report", "err", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+		}
+		return
 	}
 
 	// return a json description of all the checks
 	jsonAPIResponseWithStatus(c, checks, "checks", status)
 }
+
+func writeTextTo(w io.Writer, checks []presenters.Check) error {
+	slices.SortFunc(checks, func(a, b presenters.Check) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
+	for _, ch := range checks {
+		status := "?"
+		switch ch.Status {
+		case HealthStatusPassing:
+			status = "-"
+		case HealthStatusFailing:
+			status = "!"
+		}
+		if _, err := fmt.Fprintf(w, "%s%s\n", status, ch.Name); err != nil {
+			return err
+		}
+		if ch.Output != "" {
+			if _, err := fmt.Fprintf(newLinePrefixWriter(w, "\t"), "\t%s", ch.Output); err != nil {
+				return err
+			}
+			if _, err := fmt.Fprintln(w); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+type checkNode struct {
+	Name   string // full
+	Status string
+	Output string
+
+	Subs checkTree
+}
+
+type checkTree map[string]checkNode
+
+func newCheckTree(checks []presenters.Check) checkTree {
+	if !slices.IsSortedFunc(checks, func(a, b presenters.Check) int {
```

Comment:
why assert that the caller has to sort `checks` beforehand? Could we just sort here?

Reply:
Because we sort for the other cases and I wanted to avoid double sorting and not leave things too brittle. We could just sort in the three separate places instead. That is probably simpler. Can make it a one-liner too, instead of three.

Reply:
Updated these to: `slices.SortFunc(checks, presenters.CmpCheckName)`